addon/components/gavant-table.ts

```diff
@@ -0,0 +1,532 @@
+    //set this to `null` for fixed height/scrollable tables
+    containerSelector: string = this.elementId;
+    widthConstraint: string = 'lte-container';
+    fillMode: string = 'first-column';
+    fillColumnIndex: number | null = null;
+    resizeMode: string = 'standard';
+    bufferSize: number = 0;
+    estimateRowHeight: number = 30;
+    renderAllRows: boolean = false;
+    constrainColumnsToFit: boolean = true;
+    showEmptyFooter: boolean = false;
+
+    //component state
+    columns: NativeArray<TableColumn> = A();
+    visibleColumns: NativeArray<TableColumn> = A();
+    columnPanPosition: number = 0;
+    rows: any[] = [];
+    sorts: any[] = [];
+    isLoading: boolean = true;
+    hasMoreRows: boolean = false;
+    hasHiddenOverflow: boolean = false;
+    containerWidth: number | null = null;
+
+    //readonly attributes
+
+    /**
+     * CP that returns the default no-results text
+     *
+     * @type {string} defaultNoResultsText
+     * @memberof GavantTableComponent
+     */
+    @readOnly('noResultsText') defaultNoResultsText!: string;
+
+    /**
+     * CP that returns whether the table rows are clickable
+     *
+     * @type {boolean} clickableRows
+     * @memberof GavantTableComponent
+     */
+    @or('onRowClick', 'onRowDoubleClick') clickableRows!: boolean;
+
+    /**
+     * CP that returns whether the table is empty
+     *
+     * @type {boolean} noRows
+     * @memberof GavantTableComponent
+     */
+    @empty('rows') noRows!: boolean;
+
+    /**
+     * CP that returns whether the table is NOT loading
+     *
+     * @type {boolean} notLoading
+     * @memberof GavantTableComponent
+     */
+    @not('isLoading') notLoading!: boolean;
+
+    /**
+     * CP that returns whether a fully-loaded table is empty
+     *
+     * @type {boolean} isEmpty
+     * @memberof GavantTableComponent
+     */
+    @and('noRows', 'notLoading') isEmpty!: boolean;
+
+    /**
+     * CP that returns whether there are non-rendered columns
+     *
+     * @type {boolean} hasHiddenColumns
+     * @memberof GavantTableComponent
+     */
+    @lt('visibleColumns.length', 'columns.length') hasHiddenColumns!: boolean;
+
+    /**
+     * CP that returns whether the table is currently panned
+     *
+     * @type {boolean} isColumnsPanned
+     * @memberof GavantTableComponent
+     */
+    @gt('columnPanPosition', 0) isColumnsPanned!: boolean;
+
+    /**
+     * CP that returns true when `containerWidth` >= `minFixedColTableWidth`
+     *
+     * @type {boolean} allowFixedCols
+     * @memberof GavantTableComponent
+     */
+    @gte('containerWidth', 'minFixedColTableWidth') allowFixedCols!: boolean;
+
+    /**
+     * CP that returns which column should be used as the first
+     * visible column
+     *
+     * @type {TableColumn} firstVisibleColumn
+     * @memberof GavantTableComponent
+     */
+    @conditional('allowFixedCols', 'firstVisibleNonFixedColumn', 'visibleColumns.firstObject')
+    firstVisibleColumn!: TableColumn;
```

Comment:
Is it possible for this object to not be defined? A table with 0 columns? Or do we prevent that elsewhere?

Reply:
i dont remember that being actively prevented anywhere, so maybe to be safe we should take out this `!`, and then add if blocks/guards in places where we are assuming this to have a value.

Reply:
The `visibleColumns` array can have 0 columns technically if something goes wrong render-wise and the table-container has `width` less than the first column's `staticWidth`.